pkg/virt-controller/watch/workload-updater/workload-updater.go

```diff
@@ -0,0 +1,554 @@
+		}
+	}
+
+	c.queue.Add(key)
+}
+
+func (c *WorkloadUpdateController) deleteMigration(obj interface{}) {
+	key, err := c.getKubeVirtKey()
+	if key == "" || err != nil {
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+func (c *WorkloadUpdateController) updateMigration(old, curr interface{}) {
+	key, err := c.getKubeVirtKey()
+	if key == "" || err != nil {
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+func (c *WorkloadUpdateController) addKubeVirt(obj interface{}) {
+	c.enqueueKubeVirt(obj)
+}
+
+func (c *WorkloadUpdateController) deleteKubeVirt(obj interface{}) {
+	c.enqueueKubeVirt(obj)
+}
+
+func (c *WorkloadUpdateController) updateKubeVirt(old, curr interface{}) {
+	c.enqueueKubeVirt(curr)
+}
+
+func (c *WorkloadUpdateController) enqueueKubeVirt(obj interface{}) {
+	logger := log.Log
+	kv := obj.(*virtv1.KubeVirt)
+	key, err := controller.KeyFunc(kv)
+	if err != nil {
+		logger.Object(kv).Reason(err).Error("Failed to extract key from KubeVirt.")
+	}
+	c.queue.Add(key)
+}
+
+// Run runs the passed in NodeController.
+func (c *WorkloadUpdateController) Run(stopCh <-chan struct{}) {
+	defer controller.HandlePanic()
+	defer c.queue.ShutDown()
+	log.Log.Info("Starting workload update controller.")
+
+	// This is hardcoded because there's no reason to make thread count
+	// configurable. The queue keys off the KubeVirt install object, and
+	// there can only be a single one of these in a cluster at a time.
+	threadiness := 1
+
+	// Wait for cache sync before we start the controller
+	cache.WaitForCacheSync(stopCh, c.migrationInformer.HasSynced, c.vmiInformer.HasSynced, c.kubeVirtInformer.HasSynced)
+
+	// Start the actual work
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	log.Log.Info("Stopping workload update controller.")
+}
+
+func (c *WorkloadUpdateController) runWorker() {
+	for c.Execute() {
+	}
+}
+
+func (c *WorkloadUpdateController) Execute() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	err := c.execute(key.(string))
+
+	if err != nil {
+		log.Log.Reason(err).Infof("reenqueuing workload updates for KubeVirt %v", key)
+		c.queue.AddRateLimited(key)
+	} else {
+		log.Log.V(4).Infof("processed workload updates for KubeVirt %v", key)
+		c.queue.Forget(key)
+	}
+	return true
+}
+
+func (c *WorkloadUpdateController) isOutdated(vmi *virtv1.VirtualMachineInstance) bool {
+	if vmi.IsFinal() {
+		return false
+	}
+
+	// if the launcher image isn't detected yet, that means
+	// we don't know what the launcher image is yet.
+	// This could be due to a migration, or the VMI is still
+	// initializing. virt-controller will set it for us once
+	// either the VMI is either running or done migrating.
+	if vmi.Labels == nil {
+		return false
+	}
+
+	_, labelExists := vmi.Labels[virtv1.OutdatedLauncherImageLabel]
+	if labelExists {
+		return true
+	}
+
+	return false
+}
+
+func isMigratable(vmi *virtv1.VirtualMachineInstance) bool {
```

Comment:
Don't we have a common method for that yet? Maybe work having a common place for this?

Reply:
oh, we do actually. good catch :+1: